staking-pool-shares/src/lib.rs

```diff
@@ -0,0 +1,460 @@
+    /// Specify which account can change the staking key and the initial staking key with ED25519 curve.
+    #[init]
+    pub fn new(owner_id: AccountId, stake_public_key: Base58PublicKey) -> Self {
+        assert!(!env::state_exists(), "Already initialized");
+        Self {
+            owner_id,
+            stake_public_key: stake_public_key
+                .try_into()
+                .expect("invalid staking public key"),
+            last_epoch_height: env::epoch_height(),
+            last_locked_account_balance: 0,
+            last_account_balance: env::account_balance(),
+            total_staked_balance: 0,
+            total_staked_shares: 0,
+            accounts: Map::new(b"u".to_vec()),
+        }
+    }
+
+    /// Call to update state after epoch switched.
+    pub fn ping(&mut self) {
+        // Checking if we need there are rewards to distribute.
+        let epoch_height = env::epoch_height();
+        if self.last_epoch_height == epoch_height {
+            return;
+        }
+        self.last_epoch_height = epoch_height;
+
+        let total_balance =
+            env::account_locked_balance() + env::account_balance() - env::attached_deposit();
+        let last_total_balance = self.last_account_balance + self.last_locked_account_balance;
+
+        let total_reward = total_balance.saturating_sub(last_total_balance);
+        // The total stake is increased by both staked and unstaked rewards.
+        self.total_staked_balance += total_reward;
+
+        self.last_locked_account_balance = env::account_locked_balance();
+        self.last_account_balance = env::account_balance();
+    }
+
+    /// Call to deposit money.
+    #[payable]
+    pub fn deposit(&mut self) {
+        self.ping();
+
+        let account_id = env::predecessor_account_id();
+        let mut account = self.get_account(&account_id);
+        account.unstaked += env::attached_deposit();
+        self.save_account(&account_id, account);
+
+        if self.total_staked_balance > self.last_locked_account_balance {
+            self.restake();
+        }
+    }
+
+    /// Withdraws the non staked balance for given account.
+    pub fn withdraw(&mut self, amount: U128) {
+        self.ping();
+
+        let amount = amount.into();
+        assert!(amount > 0, "Withdrawal amount should be positive");
+
+        let account_id = env::predecessor_account_id();
+        let mut account = self.get_account(&account_id);
+        assert!(
+            account.unstaked >= amount,
+            "Not enough unstaked balance to withdraw"
+        );
+        assert!(
+            account.unstaked_available_epoch_height <= env::epoch_height(),
+            "The unstaked balance is not yet available due to unstaking delay"
+        );
+        account.unstaked -= amount;
+        self.save_account(&account_id, account);
+        Promise::new(account_id).transfer(amount);
+
+        if self.total_staked_balance > self.last_locked_account_balance {
```

Comment:
it's just weird that it's here and in `stake` given these variables only get updated in `ping`

Reply:
This is to address the issue when someone unstaked some amount in `T-3` with total stake of `X`. Now at epoch `T` the locked balance is equal to `X`, but the desired amount post 3 epochs of rewards is `X + Rewards`. Which means we have to restake `X + Rewards` to not miss out on the rewards.